bootstrap/kubeadm/internal/cloudinit/controlplane_join.go

```diff
@@ -17,12 +17,44 @@ limitations under the License.
 package cloudinit
 
 import (
-	"github.com/pkg/errors"
+	"fmt"
 
+	"github.com/pkg/errors"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha3"
 	"sigs.k8s.io/cluster-api/util/secret"
 )
 
 const (
+	controlPlaneJoinScript = `#!/bin/bash
+# Copyright 2020 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+function retry-until-working() {
+  # In case of failure, retry.
+  until $@; do
+    sleep 15
+    echo "Couldn't run command. Retrying..."
+  done
+}
+
+kubeadm join --config /tmp/kubeadm-controlplane-join-config.yaml {{.KubeadmVerbosity}}
+KUBEADM_RETURN=$?
+if [ ${KUBEADM_RETURN} -ne 0 ]; then
+  retry-until-working kubeadm join phase control-plane-join mark-control-plane --config /tmp/kubeadm-controlplane-join-config.yaml {{.KubeadmVerbosity}}
```

Comment:
If we get a nonzero exit, I don't believe there's any guarantee that it got "far enough" such that the only steps left are marking and updating status. If there was a wholesale failure (couldn't find/load the images needed, for example), the nonzero exit is a hard stop, vs. one where there was a blip talking to etcd...